backend/copr_backend/storage.py

```diff
@@ -0,0 +1,194 @@
+"""
+Support for various data storages, e.g. results directory on backend, Pulp, etc.
+"""
+
+import os
+from copr_backend.helpers import call_copr_repo
+from copr_backend.pulp import PulpClient
+
+
+def storage_for_job(job, opts, log):
+    """
+    Return an appropriate storage object for a given job
+    """
+    if job.pulp:
+        return PulpStorage(opts, log)
+    return BackendStorage(opts, log)
+
+
+class Storage:
+    """
+    Storage agnostic, high-level interface for storing and acessing our data
+    """
+
+    def __init__(self, opts, log):
+        self.opts = opts
+        self.log = log
+
+    def init_project(self, job):
+        """
+        Make sure users can enable a DNF repository for this project/chroot
+        """
```

Comment:
> it doesn't seem quite obvious what is Job

The `Job` is `copr_backend.job.BuildJob`. There is one workaround in this code where I am passing a Munch instead of the `BuildJob` because the function is called from an action and there is no `BuildJob` instance for it.

> Is't it better to accept particular options here than a Munch?

There were few reasons. Some of the methods would take quite a lot of parameters and all of them from the same object. It would look something like this:

```python
storage.publish_repository(self.job.project_owner, self.job.project_name, 
                           self.job.chroot, self.job.target_dir_name, 
                           self.job.uses_devel_repo, self.job.appstream)
```

Also, we wanted the interface to be general so that other storages besides `BackendStorage` and `PulpStorage` can be added. I didn't want to predict what parameters other storages might need.

Reply:
> I didn't want to predict what parameters other storages might need.

This is exactly the "API" design problem that I had in mind.  When you add a new option, you are forced to fix all the implementations of the interface at the same time (this is good). Or that it is much harder to add a new implementation for the interface that can be changed so easily (new attribute of munch).  So my personal preference would be to use the verbose variant :-) but ... no hard push on this.